frontend/components/markup-modal.js

```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Modal from 'react-modal';
+import ClickToSelect from '@mapbox/react-click-to-select';
+import { resolveUri } from './badge-examples';
+
+export default class MarkupModal extends React.Component {
+  constructor (props) {
+    super(props);
+    this.state = {
+      isOpen: false,
+      badgeUri: null,
+      link: null,
+      style: 'flat',
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { example } = nextProps;
+    if (example === null) {
+      this.setState({
+        isOpen: false,
+        badgeUri: null,
+      });
+    } else {
+      this.setState({
+        isOpen: true,
+        badgeUri: resolveUri(example.exampleUri || example.previewUri, nextProps.baseUri),
+        link: example.link,
+      });
+    }
+  }
+
+  generateMarkup() {
+    if (this.props.example === null) {
+      return {};
+    }
+
+    const { title } = this.props.example;
+    const { link, style } = this.state;
+
+    let badgeUri = this.state.badgeUri;
+    if (style !== 'flat') { // Default style doesn't need to be specified.
+      badgeUri += badgeUri.includes('?') ? `&style=${style}` : `?style=${style}`;
+    }
+
+    return {
+      markdown: `[![${title}](${badgeUri})](${link})`,
+      reStructuredText: `.. image:: ${badgeUri}   :target: ${link}`,
+      asciiDoc: `image:${badgeUri}[${title}]`,
+    };
+  }
+
+  renderDocumentation() {
+    const { documentation } = this.props.example || {};
+
+    return documentation ? (
+      <div>
+        <h4>Documentation</h4>
+        <div dangerouslySetInnerHTML={{ __html: documentation }} />
+      </div>
+    ) : null;
+  }
+
+  render() {
+    const { markdown, reStructuredText, asciiDoc } = this.generateMarkup();
+
+    return (
+      <Modal
+        isOpen={this.state.isOpen}
+        onRequestClose={this.props.onRequestClose}
+        contentLabel="Example Modal">
+        <form action="">
+          <p>
+            <img src={this.state.badgeUri} style={{ minHeight: '20px' }} />
+          </p>
+          <p>
+            <label>Link&nbsp;
+              <input
+                type="url"
+                value={this.state.link}
+                onChange={event => { this.setState({ link: event.target.value }); }} />
```

Comment:
This is a good example of a controlled component. You should update `dynamic-badge-maker.js` to do it this way too :)